[3.x] Improve responsiveness on underpowered Android devices

RandomShaper opened the pull request:
This PR adds a new project setting to switch to a different way of delivering input events which can greatly improve the responsiveness of games in devices that need to run multiple physics frames per idle frame (and, specially, where the input is handled in the physics frame so that the pace at which the input is taken isn't limited by the visual frame rate).

Before adding that and making it work, some preliminar steps have been taken, and are in separate commits.

Notes:
- This has helped a lot with my game. On a device that can keep 60 FPS it was fine, but when running it on an older, mid-range phone that achieves between 20-30 FPS, the input felt very laggy and the game looked more difficult and unresponsive that it actually is. **After this change (when the setting enabled), the performance is the same, but the game feels waaay more responsive.**
- If this is accepted, I'll port it to 4.0 (and probably also to iOS). I've needed to do 3.2 first since my game needed this and it wanted to have it ready as soon as possible.
- **UPDATE:** I haven't measured analytically, but I can perceive that my game is now free from some minor few millisecond hiccups that were happening once in a while. I was attributing them to other causes, but with buffering active they seem to be gone. I believe that consuming the input events from Android as they come is smoothing things out.

----------------------------
Summary of commits:

**UPDATE:** Commit structure modified.
**UPDATE:** Input handling extracted from Android OS.

### 1. Improve input event accumulation
- API has been simplified: all events now go through `parse_input_event()`. Whether they are accumulated or not depends on the `use_accumulated_input` flag.
- Event accumulation is now thread-safe (it was not needed so far, but it prepares the ground for the following changes).
- Touch drag events now support accumulation.

### 2. Add input buffering framework
Input buffering is implicitly used by event accumulation, but this commit makes it more generic so it can be enabled for other uses.

For desktop OSs it's currently not feasible given main and UI threads are the same).

### 3. Add project setting for agile input event flushing
If enabled, key/touch/joystick events will be flushed just before every idle and physics frame.

Enabling this can greatly improve the responsiveness to input, specially in devices that need to run multiple physics frames per each idle frame, because of not being powerful enough to run at the target frame rate.

This will only work for platforms using input buffering (regardless event accumulation). Currenly, only Android does so, but could be implemented for iOS in an upcoming PR.

### 4. Switch to input buffering on Android

Switch to input buffering on Android

Key, touch and joystick events will be passed directly from the UI thread to Godot, so they can benefit from agile input flushing.

As another consequence of this new way of passing events, less Java object are created at runtime (`Runnable`), which is good since the garbage collector needs to run less.

`AndroidInputHandler` is introduced to have a smaller cross-thread surface. `main_loop_request_go_back()` is removed in favor just inline calling `notification()` on the `MainLoop` at the most caller's convenience.

Lastly, `get_mouse_position()` and `get_mouse_button_state()` now just call through `InputDefault` to avoid the need of sync of mouse data tracked on the UI thread.

RandomShaper committed:
Improve input event accumulation

- API has been simplified: all events now go through `parse_input_event()`. Whether they are accumulated or not depends on the `use_accumulated_input` flag.
- Event accumulation is now thread-safe (it was not needed so far, but it prepares the ground for the following changes).
- Touch drag events now support accumulation.

RandomShaper committed:
Add project setting for agile input event flushing

If enabled, key/touch/joystick events will be flushed just before every idle and physics frame.

Enabling this can greatly improve the responsiveness to input, specially in devices that need to run multiple physics frames per each idle frame, because of not being powerful enough to run at the target frame rate.

This will only work for platforms using input buffering (regardless event accumulation). Currenly, only Android does so, but could be implemented for iOS in an upcoming PR.

pouleyKetchoupp commented on core/os/input_event.cpp:
```diff
@@ -1000,6 +1000,23 @@ String InputEventScreenDrag::as_text() const {
+
+	Ref<InputEventScreenDrag> drag = p_event;
+	if (drag.is_null())
+		return false;
```
Brackets are missing here.

RandomShaper replied:
Here I just imitated `InputEventMouseMove::accumulate()`. Since in Godot 3 we are not enforcing the always-use-braces style, I decided to just do the same as in that other function, for consistency.

pouleyKetchoupp reviewed (changes requested):
Nice work! It seems like a good idea. I'll try to give it a test some time this week.

RandomShaper commented on platform/android/java/lib/src/org/godotengine/godot/Godot.java:
```diff
@@ -977,7 +977,7 @@ private boolean obbIsCorrupted(String f, String main_pack_md5) {
 	public boolean gotTouchEvent(final MotionEvent event) {
 
-		final int evcount = event.getPointerCount();
+		int evcount = event.getPointerCount();
```
I noticed most `final`s where there so that the `Runnable` classes could access them. That's a requirement in Java for a class to access the scope of the function it's declared in.

I started removing them, but eventually I thought that they could very well stay anyway. But I forgot to restore this one.

RandomShaper commented on platform/android/java_godot_lib_jni.cpp:
```diff
@@ -50,14 +50,15 @@
 
 static bool initialized = false;
-static int step = 0;
+static std::atomic<int> step; // Shared between UI and render threads
```
Hmm... I'll give this a deeper thought.

RandomShaper replied:
I get it. Good point. Done!

Calinou commented:
As for event accumulation, do we need it for input events other than InputEventMouseMotion and touch drag events? Most people can't press keys or gamepad buttons more than ~15 times in a second :slightly_smiling_face:

RandomShaper commented:
The only other kind of event that may be accumulated is joystick axis movement, but I think that would introduce problems, like when you want to parse a stick sequence for a fighting game. Accumulation there wouldn't be as easy as keeping the most recent values. Therefore, I'd rather keep accumulation limited to pointer-like moves.

In addition, just in case I'll clarify that the only thing this PR does to accumulation is enabling it for touch. The point of this PR is getting more chances of having the most recent input when code checking it is about to run and that is orthogonal to accumulation.

RandomShaper commented:
I haven't measured analytically, but I can perceive that my game is now free from some minor few millisecond hiccups that were happening once in a while. I was attributing them to other causes, but with buffering active they seem to be gone. I believe that consuming the input events from Android as they come is smoothing things out.

(I've added this as an additional note to the description of this PR.)

pouleyKetchoupp commented on main/main.cpp:
```diff
@@ -2069,10 +2073,14 @@ bool Main::iteration() {
-
 	for (int iters = 0; iters < advance.physics_steps; ++iters) {
 
+		if (agile_input_event_flushing) {
```
Should `InputDefault::get_singleton()->is_using_input_buffering()` be used here too to avoid possible side effects on platforms other than Android, since the documentation says it's implemented only on Android?

RandomShaper replied:
As far as my brain gets, it shouldn't be needed, because the buffer would just be empty. However, I think it's a good idea in any case. Let's be defensive. Pushing change in 3, 2, 1...

pouleyKetchoupp commented:
I don't have a test case to specifically check the hiccups you describe but I've just tested on Android (low-end device - Archos 55 Helium) and didn't spot any issue with or without agile input events flushing enabled.

RandomShaper commented:
Cool. Thanks for testing.

El mié., 23 sept. 2020 11:01, Camille Mohr-Daurat <notifications@github.com>
escribió:

> I don't have a test case to specifically check the hiccups you describe
> but I've just tested on Android (low-end device - Archos 55 Helium) and
> didn't spot any issue with or without agile input events flushing enabled.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/42220#issuecomment-697232847>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2AFNU4HQ4O3XZH4FVKVGTSHG2M3ANCNFSM4RT5SX5A>
> .
>

pouleyKetchoupp reviewed (approved):
Seems good to me.

It would be nice however to be able to test the issue itself. Is there a way you could provide a minimal project?

RandomShaper commented:
This will require a rebase, but I'd like to know if there's a chance it will be accepted before doing the work.

reduz commented:
In think on the general idea its ok, but I don't see why you would want to turn off event accumulation for what comes from the OS, and also you need to leave input_event alone to work without accumulation so the users can throw their custom events, so I suggest just keep accumulate_event and use this instead from the OSs.

RandomShaper commented:
This PR doesn't change how input accumulation works. The ability to enable/disable it was already there.

Therefore, at the OS "layer" you still don't know if it's enabled or not. OS feeds the events to the engine just as it did formerly, except that happens in a way that is also compatible with the input buffering this PR adds.

RandomShaper commented:
Also note that this PR, being for 3.2, adds input buffering as an option to avoid changing behavior of current projects. For 4.0 it could be just the builtin behavior in every platform supporting it, to avoid unneeded complexity.

m4gr3d reviewed (commented):
@RandomShaper Can you clarify why the engine can now handles input directly from the UI thread?
For example there are several fields in `os_android` (e.g: `hover_prev_pos`) which after this change will be updated from the UI thread but read from the render thread.

RandomShaper commented:
@m4gr3d, I'm failing to see what you're pointing out about `hover_prev_pos`.

Members of `OS_Android` like this are constructed on the UI thread, before the render thread is created and then, read/written on the render thread exclusively:
```
// Called on the UI thread
void OS_Android::process_hover(int p_type, Point2 p_pos) {
	// https://developer.android.com/reference/android/view/MotionEvent.html#ACTION_HOVER_ENTER
	switch (p_type) {
		case 7: // hover move
		case 9: // hover enter
		case 10: { // hover exit
			Ref<InputEventMouseMotion> ev;
			ev.instance();
			ev->set_position(p_pos);
			ev->set_global_position(p_pos);
			ev->set_relative(p_pos - hover_prev_pos);
			input->parse_input_event(ev);
			hover_prev_pos = p_pos;
		} break;
	}
}
```

Can you elaborate or point me to another case? Maybe there's some confusion because of the pending rebase?

m4gr3d commented:
@RandomShaper There may be a miscommunication; you mention there are `read/written on the render thread exclusively` but in the example you gave, there is a comment above the method stating that it's called on the UI thread, suggesting the fields are being written on the UI thread and read from the render thread (sing the engine loop is being driven by the render thread).

As additional example, you update the `step` variable in `java_godot_lib_jni.cpp` to be atomic to avoid a similar issue:
```
static std::atomic<int> step; // Shared between UI and render threads
```

With the current approach in the PR, the same would need to be done for all the fields in `OS_Android` to make them safe for access. Let me know if I should clarify more.

Calinou commented:
PS: This PR is likely tangentially related to [this proposal](https://github.com/godotengine/godot-proposals/issues/1288).

RandomShaper commented:
@m4gr3d, I was wrong in my comment. I checked the code too early in the morning. The case of `hover_prev_pos` is actually easy, because it's only used on the UI thread.

I think `step` is the only one that required some thread safety mechanism.

But I may be wrong. Please let me know.

m4gr3d commented:
@RandomShaper Looks like this needs to be rebased.

akien-mga commented on main/main.cpp:
```diff
@@ -2086,10 +2090,14 @@ bool Main::iteration() {
+			InputDefault::get_singleton()->flush_buffered_events();
+		}
+
+		Engine::get_singleton()->_in_physics = true;
```
Are there any risks tied to toggling this on/off for each physics iteration like this?

RandomShaper replied:
Just to decide one way or the other (since I don't see a strong reason for either), I decided this, so the input callbacks are still called with `Engine::is_in_physics_frame()` being `false`, to make it a bit more transparent to the game.

I can't see any potential risks.

pouleyKetchoupp replied:
Yeah, it's hard to predict potential side effects, but it makes sense to keep input events reported as not being part of the physics tick, and we'll see if it causes issues in some edge cases by testing over time.

RandomShaper commented:
Rebased and re-tested the touch part.

It'd be nice that someone tests that Android mouse still works (with and without agile flushing enabled).

akien-mga commented:
CC @thebestnom who has been doing a lot of work on Android input.

RandomShaper commented:
I've force-pushed to ammend the first commit, so that `use_accumulated_input` is initialized to `false`. The point of that is that, whether accumulation is used depends on the specific OS abstractions having opted-in for it. When it was `true`, any OS that wasn't opting-in for accumulation and hadn't support for input buffering would not react at all to touch.

A silly but relevant mistake that I found by testing this on iOS.

RandomShaper commented:
Given there's no much hope at this point that this will get more reviews, I've rebased with the hope that it can be merged at some point, maybe by the next alpha/beta so it's usert-tested.

RandomShaper committed:
Add input buffering framework

Input buffering is implicitly used by event accumulation, but this commit makes it more generic so it can be enabled for other uses.

For desktop OSs it's currently not feasible given main and UI threads are the same).

RandomShaper commented:
@m4gr3d, I've done what you suggested on IRC. Please check, especially the 4th commit,

I haven't re-tested it yet after the latest changes. I'll do soon.

m4gr3d commented on platform/android/java_godot_lib_jni.cpp:
```diff
@@ -215,20 +221,13 @@ JNIEXPORT void JNICALL Java_org_godotengine_godot_GodotLib_newcontext(JNIEnv *en
 	}
 }
 
-JNIEXPORT void JNICALL Java_org_godotengine_godot_GodotLib_back(JNIEnv *env, jclass clazz) {
-	if (step == 0)
-		return;
-
-	os_android->main_loop_request_go_back();
-}
-
```
This seems to still be used in [`Godot.java`](https://github.com/godotengine/godot/blob/2269c2ebe2cd9eee16bb685cb945d67991d4b27b/platform/android/java/lib/src/org/godotengine/godot/Godot.java#L941).

RandomShaper replied:
I failed to find the call. Will restore it.

RandomShaper replied:
Done.

m4gr3d commented on platform/android/os_android.cpp:
```diff
@@ -246,11 +246,11 @@ bool OS_Android::is_mouse_grab_enabled() const {
 
 Point2 OS_Android::get_mouse_position() const {
-	return hover_prev_pos;
+	return input->get_mouse_position();
 }
 
 int OS_Android::get_mouse_button_state() const {
-	return buttons_state;
+	return input->get_mouse_button_mask();
 }
```
How are the mouse position and the mouse button mask updated?

RandomShaper replied:
`InputDefault` tracks them as well, and in a more comprehensive way. My idea is to keep `AndroidInputHandler`s own tracking by now, and in the future, when we remove OS-specific tracking from everywhere, we bring that last amount of sanity here.

m4gr3d commented on platform/android/os_android.h:
```diff
@@ -200,24 +158,72 @@ class OS_Android : public OS_Unix {
 	~OS_Android();
 };
 
+// This class encapsulates all the handling of input events that come from the Android UI thread.
```
Looks good! Let's move the class into its own header and source files.
Can you also add a warning to the class description that all the class fields and methods should remain thread safe.

RandomShaper replied:
Will do.

RandomShaper replied:
Done.

m4gr3d reviewed (changes requested):
The commit looks pretty good. I've added some minor feedback and comments.

m4gr3d commented on main/main.cpp:
```diff
@@ -1406,6 +1406,8 @@ Error Main::setup2(Thread::ID p_main_tid_override) {
 
 	InputDefault *id = Object::cast_to<InputDefault>(Input::get_singleton());
 	if (id) {
+		agile_input_event_flushing = GLOBAL_DEF("input_devices/buffering/agile_event_flushing", false);
+
```
Shouldn't this property be used to update the `InputDefault` singleton using `InputDefault::set_use_input_buffering(...)`?
If not, what's the intended use for the method?

m4gr3d commented on platform/android/os_android.h:
```diff
@@ -200,24 +158,72 @@ class OS_Android : public OS_Unix {
-	void process_event(Ref<InputEvent> p_event);
 	void init_video_mode(int p_video_width, int p_video_height);
 
 	virtual bool is_joy_known(int p_device);
 	virtual String get_joy_guid(int p_device) const;
```
Should these methods also be moved to the `AndroidInputHandler` class?

RandomShaper replied:
My criteria for moving functions to `AndroidInputHandler` has been the following:
- They were not overrides of `OS`'s functions, since those are only guaranteed to be safe on the Android render thread (Godot main thread) and therefore don't need an special treatment.
- They will be called from the Android UI thread.
- They only call thread-safe functions.
All the points have turned out to be met at the same time for all the functions of interest.

RandomShaper committed:
Switch to input buffering on Android

Key, touch and joystick events will be passed directly from the UI thread to Godot, so they can benefit from agile input flushing.

As another consequence of this new way of passing events, less Java object are created at runtime (`Runnable`), which is good since the garbage collector needs to run less.

`AndroidInputHandler` is introduced to have a smaller cross-thread surface. `main_loop_request_go_back()` is removed in favor just inline calling `notification()` on the `MainLoop` at the most caller's convenience.

Lastly, `get_mouse_position()` and `get_mouse_button_state()` now just call through `InputDefault` to avoid the need of sync of mouse data tracked on the UI thread.

RandomShaper commented:
Rebased and re-tested for no regressions (mouse & keyboard on Windows and touch on Android and iOS).

m4gr3d reviewed (approved):
The update looks good! I've added one more feedback, once that's resolved, this should be good to merge!

RandomShaper commented on platform/android/java_godot_lib_jni.cpp:
```diff
@@ -237,6 +244,7 @@ JNIEXPORT void JNICALL Java_org_godotengine_godot_GodotLib_step(JNIEnv *env, jcl
 		// Since Godot is initialized on the UI thread, _main_thread_id was set to that thread's id,
 		// but for Godot purposes, the main thread is the one running the game loop
 		Main::setup2(Thread::get_caller_id());
+		input_handler = new AndroidInputHandler();
```
Makes sense. However, I first tried to construct the input handler at `Java_org_godotengine_godot_GodotLib_initialize()`. However, it seems that's too soon for `InputDefault` to be created and is causing a flood of errors and the game is unable to start. Regarding the fact it's created on the render thread and used from the UI one, shouldn't be a problem, since the `step` variable issues memory barriers on increment and comparison. Not that I like that approach much, though, so if you can suggest a good place to move the construction of the input handler to, I'd listen to you.

m4gr3d replied:
Good point. Looking at the previous logic, `InputDefault` was initialized for `OS_Android` in `OS_Android::initialize(...)` which itself is called in `Main::setup2(...)`, so where you placed the init logic is actually correct.

The only alternative I can think of is to not store a reference to `InputDefault` in the class and always just use `InputDefault::get_singleton()` instead, but that seems overly bothersome..

I'll leave the decision to you, the current location is fine by me.

RandomShaper replied:
Alright then. Let's leave it this way. We can always improve it later; this is likely to be revisited.
